Initial docs pass: adding docs sections and adding model and signal API docs

DanPuzzuoli opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

Initial pass of docs development.

General additions:
- Added "How-to guides" folder/category to documentation (currently empty)
- Added "Discussions" folder/category to documentation (currently empty)

Specific docs modifications:
- Signals API documentation
    - Translated `docs/tutorials/Signals.ipynb` into API docs (and deleted the notebook).
    - Still need to handle transfer functions, but can be a separate PR.
- Models API documentation
    - Added documentation for different model types.
    - Included section on `evaluation_mode`s, so deleted `docs/tutorials/evaluation_mode_demo.ipynb`

DanPuzzuoli commented:
Some notes for reviewers:
- For the API references, there are various table of contents being automatically generated in various places; I'm not sure what the right things to display are. Right now they're showing the subheadings of the module-level documentation, but it doesn't seem quite right.
- The `solvers` section of the API references is pretty old and I want to simplify it, but am planning on doing it in a different PR which modifies the relevant functions, so don't worry about the specifics of this particular folder's documentation - I'll make it conform later based on whatever we decide here.

chriseclectic commented on qiskit_dynamics/signals/__init__.py:
```diff
@@ -12,14 +12,195 @@
+DiscreteSignal
+^^^^^^^^^^^^^^
+
+A :class:`~qiskit_dynamics.signals.DiscreteSignal`  is specified
```
```suggestion
A :class:`~qiskit_dynamics.signals.DiscreteSignal`  is a signal specified
```

chriseclectic commented on qiskit_dynamics/signals/__init__.py:
```diff
@@ -12,14 +12,195 @@
+^^^^^^^^^^^^^^
+
+A :class:`~qiskit_dynamics.signals.DiscreteSignal`  is specified
+as an array of samples with a sample width and an optional start point (defaults to ``0.``).
```
I think this needs more description what sample width is (and should mention `dt`). Something like:

A discrete signal is specified by an Array [s_0, ..., s_k] and sample width `dt` and is evaluated as s(t) = s_{floor(t/dt)} (or something like that). By default discrete signals are defined to start at t=0, and hence have duration n * dt, where n is the number of samples, but a custom start time can be set by using the <whatever kwarg is> kwarg.

DanPuzzuoli replied:
Just put a version of this in.

chriseclectic commented on qiskit_dynamics/signals/__init__.py:
```diff
@@ -12,14 +12,195 @@
+
+A :class:`~qiskit_dynamics.signals.DiscreteSignal`  is specified
+as an array of samples with a sample width and an optional start point (defaults to ``0.``).
+:class:`~qiskit_dynamics.signals.DiscreteSignal` inherits the same API as
```
Don't think it is necessary to mention this, it should be clear, or mentioned at the start of the section on signals that all signals have same basic API.

DanPuzzuoli replied:
Added this at the start, along with a table listing all signal class and giving a single sentence description of what they are supposed to represent.

chriseclectic commented on qiskit_dynamics/signals/__init__.py:
```diff
@@ -12,14 +12,195 @@
+A :class:`~qiskit_dynamics.signals.DiscreteSignal`  is specified
+as an array of samples with a sample width and an optional start point (defaults to ``0.``).
+:class:`~qiskit_dynamics.signals.DiscreteSignal` inherits the same API as
+:class:`~qiskit_dynamics.signals.Signal`, and once instantiated can be interacted with
```
Unnecessary sentence

DanPuzzuoli replied:
Removed.

chriseclectic commented on qiskit_dynamics/signals/__init__.py:
```diff
@@ -12,14 +12,195 @@
+:class:`~qiskit_dynamics.signals.Signal`, and once instantiated can be interacted with
+in the same way.
+
+.. jupyter-execute::
```
I don't think this execute block really adds anything, especially since all the used kwargs haven't been defined or mentioned. It's probably more confusing than helpful here.

DanPuzzuoli replied:
I've just turned this into a single line showing a discrete signal being instantiated.

chriseclectic commented on qiskit_dynamics/signals/__init__.py:
```diff
@@ -12,14 +12,195 @@
+    discrete_signal.draw(t0=0., tf=1.5, n=100,
+                         function='complex_value', axis=axs[1], title='Complex value')
+
+Note that the carrier wave is still analog for :class:`~qiskit_dynamics.signals.DiscreteSignal`.
```
What does this mean?

DanPuzzuoli replied:
I've removed this as I agree that it's confusing, and what I meant to say is already implied by the general definition.

chriseclectic commented on qiskit_dynamics/signals/__init__.py:
```diff
@@ -12,14 +12,195 @@
+    signal_sum = signal1 + signal2
+
+
+:class:`~qiskit_dynamics.signals.SignalSum` again implements the same API
```
Mention at start of signal section rather then mentioning each time. Actually there should probably a section on the base API for signals that all subclasses implement.

DanPuzzuoli replied:
I've added a discussion of this API at the beginning, along with a table listing the names of each signal class and a very broad signle sentence description of what it represents.

chriseclectic commented on qiskit_dynamics/signals/__init__.py:
```diff
@@ -12,14 +12,195 @@
+so that the sum can be evaluated using purely array-based operations.
+
+Signal multiplication is also supported via :class:`~qiskit_dynamics.signals.SignalSum`
+and :class:`~qiskit_dynamics.signals.DiscreteSignalSum` using the identity:
```
There should be a summary table somewhere (probably at the start) of this doc that lists all the signal classes with a short description of what they represent.

DanPuzzuoli replied:
See previous comment.

chriseclectic reviewed (changes requested):
Good changes. Some general recommendations to improve readability are:
- Always start class or function doc strings with a short sentence description, followed by blank line then longer paragraphs of description if required (PEP8). The first line is what is shown in sphinx autosummary tables so you want it short.
- Avoid using sphinx commands (like `math`) in that first line description since they are difficult to read in source code or using `help` command in python.
- Add some more sub headings to break up the very long doc strings in some modules.

DanPuzzuoli commented on qiskit_dynamics/models/__init__.py:
```diff
@@ -12,40 +12,166 @@
+:meth:`~qiskit_dynamics.models.LindbladModel.evaluate` method will raise an error unless
+a vectorized evaluation mode is set (see below).
+
+Rotating Frames
```
Is there anything to be done here, or are you just saying it's now a sub-sub-heading?

DanPuzzuoli commented on qiskit_dynamics/models/__init__.py:
```diff
@@ -12,40 +12,166 @@
+    model.rotating_frame = frame_operator
+
+where ``frame_operator`` is specification of :math:`F=-iH`
+(either giving :math:`F` or :math:`H`) of valid type. Setting this property modifies
```
Implemented changes:
- Described initial transformation in terms of Hermitian operator `H_0` (used `H_0` instead of `H` to avoid potential confusion with the original Hamiltonian `H(t)`).
- Stated the accepted types for `frame_operator` below the python code block, including `RotatingFrame`, and point the reader to "see below" for further description about `RotatingFrame`.
- Slightly expanded the blurb about `RotatingFrame` to state that internally it works with the anti-Hermitian matrix `F`, but that it can be specified in terms of `H_0` which will automatically be converted to `F`.

DanPuzzuoli commented on qiskit_dynamics/models/__init__.py:
```diff
@@ -12,40 +12,166 @@
+
+    model.set_evaluation_mode('sparse')
+
+See ``set_evaluation_mode`` for each model class for available modes.
```
Let's maybe leave this discussion for the other PR since it's a non-docs change.

DanPuzzuoli commented on qiskit_dynamics/models/__init__.py:
```diff
@@ -12,40 +12,166 @@
+rotating frames requires more restrictive choice of frames. For example, diagonal frame
+operators exactly preserve sparsity.
+
+The different modes are organized via a series of objects called ``OperatorCollection`` s,
```
There actually isn't even an `OperatorCollection` object. I've changed:

'series of objects called ``OperatorCollection`` s,'

to 

'series of objects called "operator collections",'.

I'm intending these classes to never be used directly by users, they are purely convenience for organizing/abstracting things within the model classes. I thought they were worth a mention, but should we maybe even completely remove them from the API documentation?

DanPuzzuoli commented on qiskit_dynamics/models/operator_collections.py:
```diff
@@ -56,31 +53,18 @@ def num_operators(self) -> int:
-        as :math:`G(t)`, returns this independent representation.
-        If the model cannot be represented in such a
-        manner (c.f. Lindblad model), then errors."""
+        r"""If possible, compute a representation of :math:`\Lambda(c, \cdot)`"""
```
changed

DanPuzzuoli commented on qiskit_dynamics/models/operator_collections.py:
```diff
@@ -56,31 +53,18 @@ def num_operators(self) -> int:
-        :math:`y` provided the values of each signal
-        component :math:`s_j(t)`. Must be defined for all
-        models."""
+        r"""Compute :math:`\Lambda(c, y)`."""
```
Just changed it to a sentence.

DanPuzzuoli commented on qiskit_dynamics/models/operator_collections.py:
```diff
@@ -93,10 +77,9 @@ def copy(self):
-    need left multiplication–those of the form
-    :math:`\dot{y} = G(t)y(t)`, where :math:`G(t) = G_d + \sum_j s_j(t) G_j`.
-    Can evaluate :math:`G(t)` independently of :math:`y`.
+    r"""Concrete operator collection representing a function of the form
```
Changed.

DanPuzzuoli commented on qiskit_dynamics/models/operator_collections.py:
```diff
@@ -117,8 +100,7 @@ def num_operators(self) -> int:
     def evaluate(self, signal_values: Array) -> Array:
-        r"""Evaluates the operator :math:`G(t)` given
-        the signal values :math:`s_j(t)` as :math:`G(t) = \sum_j s_j(t)G_j`."""
+        r"""Evaluate :math:`G = G_d + \sum_j c_jG_j`."""
```
Changed to reference to the structure described in the init function doc.

DanPuzzuoli commented on qiskit_dynamics/models/operator_collections.py:
```diff
@@ -200,9 +182,10 @@ def evaluate_rhs(self, signal_values: Array, y: Array) -> Array:
 
 class DenseLindbladCollection(BaseOperatorCollection):
-    r"""Calculation object for the Lindblad equation:
+    r"""Object for evaluating:
```
added.

DanPuzzuoli commented on qiskit_dynamics/signals/__init__.py:
```diff
@@ -12,14 +12,195 @@
 .. currentmodule:: qiskit_dynamics.signals
 
-Tools for constructing time-dependent signals in RHS :mod:`qiskit_dynamics.models`.
+Signal objects represent a class of real-valued functions of the form:
```
Agreed, though didn't commit this due to the typo "model" v.s. "module". I've copied your suggestion and fixed this typo though.

DanPuzzuoli commented:
I believe I've addressed everything @chriseclectic :

- I've committed almost all of the direct code suggestions you made (aside from one which had a typo)
- I've tried to implement all changes, and have replied to every comment.
- The biggest changes came in the signals init file due to the nature of your comments. Several of the comments are pretty direct and simple, but the biggest change is a new section (based on your suggestion) at the beginning describing the common API, and providing a table with each type of signal and a brief description as to what its role is. I've also removed all of the repetitive "this implements the same API" stuff.
- I think one final question (mentioned in a comment) is whether to even include the operator collection stuff in the built docs, as it's meant for purely internal usage. I think it may actually make sense to cut it from these docs based on this. Thoughts?
